src/train.py

```diff
@@ -157,47 +241,106 @@ def optimizer_step(loss, model, optimizer, args):
     return loss.detach().cpu().item()
 
 def train_sp_step(losses, model, optimizer, batch, args):
-    asr_loss, tts_loss = supervised_step(model, batch)
-    loss = tts_loss + asr_loss
+    model.text_m.train()
+    model.speech_m.train()
+    model.discriminator.eval()
+
+    if args.use_discriminator:
+        asr_loss, tts_loss, t_d_loss, s_d_loss = supervised_step(model, batch, args.use_discriminator)
+        loss = tts_loss + asr_loss + t_d_loss + s_d_loss
+    else:
+        asr_loss, tts_loss = supervised_step(model, batch)
+        loss = tts_loss + asr_loss
 
     # Take a optimizer and append losses here!
     optimizer_step(loss, model, optimizer, args)
 
     # Log losses
     losses['asr_'].append(asr_loss.detach().cpu().item())
     losses['tts_'].append(tts_loss.detach().cpu().item())
+    if args.use_discriminator:
+        losses['t_sp_d'].append(t_d_loss.detach().cpu().item())
+        losses['s_sp_d'].append(s_d_loss.detach().cpu().item())
 
 def train_ae_step(losses, model, optimizer, batch, args):
+    model.text_m.train()
+    model.speech_m.train()
+    model.discriminator.eval()
 
-    t_ae_loss, s_ae_loss = autoencoder_step(model, batch)
-    loss = t_ae_loss + s_ae_loss
+    if args.use_discriminator:
+        t_ae_loss, s_ae_loss, t_d_loss, s_d_loss = autoencoder_step(model, batch, args.use_discriminator)
+        loss = t_ae_loss + s_ae_loss + t_d_loss + s_d_loss
+    else:
+        t_ae_loss, s_ae_loss = autoencoder_step(model, batch)
+        loss = t_ae_loss + s_ae_loss
 
     # Take a optimizer step!
     optimizer_step(loss, model, optimizer, args)
 
     # Log losses
     losses['t_ae'].append(t_ae_loss.detach().cpu().item())
     losses['s_ae'].append(s_ae_loss.detach().cpu().item())
+    if args.use_discriminator:
+        losses['t_ae_d'].append(t_d_loss.detach().cpu().item())
+        losses['s_ae_d'].append(s_d_loss.detach().cpu().item())
 
 def train_cm_step(losses, model, optimizer, batch, args):
+    model.text_m.train()
```

Comment:
For efficency sake, can we just set these once at the top of train?

Reply:
It cannot be set once at the top of train because we alternate training steps between the model and the discriminator. It can be set inside of the training loop, but this is more consistent in case we want to do some different training experiments without having to refactor the training loop again.